### ISSUE TITLE
[Validator] Encourage type safety

### ISSUE THREAD

**Issue by kiler129** [symfony/symfony-docs#9979]
Example provided at https://symfony.com/doc/current/validation/custom_constraint.html#creating-the-validator-itself shows example validation as:

```php
    public function validate($value, Constraint $constraint)
    {
        if (!preg_match('/^[a-zA-Z0-9]+$/', $value, $matches)) {
            $this->context->buildViolation($constraint->message)
                ->setParameter('{{ string }}', $value)
                ->addViolation();
        }
    }
```

This is a perfectly fine example, but sets a trap creating assumption that `$value` is always a string. I think at minimum we should include:

```php
        if (!is_string($value)) {
            throw new UnexpectedTypeException($value, 'string');
        }
```

if not even something like `DateTimeValidator` has:

```php
        if (!is_scalar($value) && !(is_object($value) && method_exists($value, '__toString'))) {
            throw new UnexpectedTypeException($value, 'string');
        }

        $value = (string) $value;
```

WDYT?

**Comment by xabbuh** [comment]
👍 from my side. I (almost) always do this in custom constraint validators. Usually, this should never be triggered, but in case you misconfigure some validation it may greatly help debugging because you end up with much more meaningful error messages.

**Comment by javiereguiluz** [comment]
@kiler129 thanks for reporting this issue. It was fixed in #9993.
